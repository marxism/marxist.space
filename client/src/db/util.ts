import * as pluralize from 'pluralize'

export const uniq = (xs) => xs.filter((v, i, s) => s.indexOf(v) === i)

export const tagAliases = [
  ['abolition', 'police', 'prison', 'cop', 'policing', 'jail'],
  ['anarchist', 'anarchism', 'anarchy'],
  ['british', 'english', 'britain', 'england', 'uk'],
  ['castro', 'fidel', 'cuba', 'che'],
  ['china', 'prc', 'deng', 'xiaoping', 'ccp', 'cpc'],
  ['covid', 'covid19', 'covid-19', 'coronavirus'],
  ['debunked', 'debunk', 'debunking', 'myth', 'lie'],
  ['deng', 'xiaoping'],
  ['glossary', 'definition', 'dictionary'],
  ['health', 'healthcare', 'health care'],
  ['hk', 'hongkong', 'hong-kong', 'hong kong'],
  ['imperialism', 'anti-imperialism', 'decolonization', 'colonization'],
  ['israel', 'palestine'],
  ['latin', 'latam'],
  ['library', 'libraries'],
  ['list', 'collection'],
  ['mao', 'zedong', 'tsetung'],
  [
    'ml',
    'marxist-leninist',
    'leninist',
    'stalinist',
    'marxist leninist',
    'leninism',
    'stalinism',
    'marxism-leninism',
    'marxism leninism',
  ],
  ['mlm', 'maoist', 'mao', 'maoism'],
  ['news', 'periodical', 'media', 'msm'],
  ['north-korea', 'dprk', 'korea', 'juche'],
  ['soviet-union', 'ussr', 'soviet'],
  ['sw', 'sex-work', 'sex work'],
  ['trostky', 'trot', 'troskyite', 'troskyism'],
  ['usa', 'us', 'america', 'united states', 'united-states'],
  ['uyghur', 'uighur', 'xinjiang'],
  ['xi', 'jinping'],
  ['zapatista', 'ezln'],
]

export const addPlurals = (tags: Array<string>): Array<string> =>
  uniq(
    // @ts-ignore
    tags.flatMap((tag) => [pluralize.plural(tag), pluralize.singular(tag)])
  )

export const addTagAliases = (tags: Array<string>): Array<string> => {
  const pluralized = addPlurals(tags)
  // @ts-ignore
  const possibleAliases = tagAliases
    .filter((xs) => xs.find((x) => pluralized.includes(x)))
    // @ts-ignore
    .flat()
  return uniq([...pluralized, ...possibleAliases])
}
