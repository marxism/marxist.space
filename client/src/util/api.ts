import { getJson as _getJson } from 'fetchyeah'

const baseUrl = '/api'

const getJson = (url) => _getJson(`${baseUrl}/${url}`)

export const fetchResources = () => getJson('all')

export const fetchTags = () => getJson('tags')

export const searchResources = (text = '') => getJson(`search?text=${text}`)

export const filterResources = (tags: Array<string> = []) =>
  getJson(`filter?tags=${tags.join(',')}`)
