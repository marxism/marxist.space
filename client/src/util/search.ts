export const getQueryParams = (search = '') => {
  const params = {}
  search
    .substring(1) // remove leading '?'
    .split('&')
    .filter((value) => value.trim().length)
    .forEach((part) => {
      const [key, value] = part.split('=')
      params[key.toLowerCase()] = decodeURIComponent(value)
    })
  return params
}

