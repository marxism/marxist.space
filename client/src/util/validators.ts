export const checkMaxLength = (value, len) =>
  value && value.length <= len
    ? undefined
    : `must be less than ${len} characters`

export const checkMinLength = (value, len) =>
  value && value.length >= len
    ? undefined
    : `must be more than ${len} characters`

export const checkValidChars = (value) =>
  /^[a-zA-Z0-9_-]+$/.test(value) ? undefined : 'contains invalid characters'

export const checkIfTrimmed = (value) =>
  value.trim() === value ? undefined : 'cannot start or end with whitespace'

export const validUrl = (value) => {
  try {
    new URL(value)
    return undefined
  } catch (error) {
    return 'must be a valid url'
  }
}

export const max = (len) => (value) => checkMaxLength(value, len)
export const min = (len) => (value) => checkMinLength(value, len)
export const validChars = (value) => checkValidChars(value)
export const trimmed = (value) => checkIfTrimmed(value)

export const required = (value) => (value ? undefined : 'required')
export const resourceType = (value) =>
  value === 'link' || value === 'text'
    ? undefined
    : 'must be link or text resource'
export const titleValidator = (value) =>
  required(value) || checkMaxLength(value, 100)
export const textResourceValidator = (value) =>
  required(value) || checkMinLength(value, 4)
export const urlValidator = (value) => required(value) || validUrl(value)
export const typeValidator = (value) => required(value) || resourceType(value)
