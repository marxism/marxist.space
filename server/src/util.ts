import * as pluralize from 'pluralize'

export const uniq = (xs) => xs.filter((v, i, s) => s.indexOf(v) === i)

export const tagAliases = [
  ['abolition', 'police', 'prison', 'cop', 'policing'],
  ['anarchist', 'anarchism', 'anarchy'],
  ['castro', 'fidel', 'cuba', 'che'],
  ['china', 'prc', 'deng', 'xiaoping'],
  ['debunked', 'debunk', 'debunking', 'myth', 'lie'],
  ['deng', 'xiaoping'],
  ['glossary', 'definition', 'dictionary'],
  ['health', 'healthcare', 'health care'],
  ['hk', 'hongkong', 'hong-kong', 'hong kong'],
  ['israel', 'palestine'],
  ['latin', 'latam'],
  ['library', 'libraries'],
  ['list', 'collection'],
  ['mao', 'zedong', 'tsetung'],
  [
    'ml',
    'marxist-leninist',
    'leninist',
    'stalinist',
    'marxist leninist',
    'leninism',
    'stalinism',
    'marxism-leninism',
    'marxism leninism',
  ],
  ['mlm', 'maoist', 'mao', 'maoism'],
  ['news', 'periodical', 'media', 'msm'],
  ['north-korea', 'dprk', 'korea', 'juche'],
  ['soviet-union', 'ussr', 'soviet'],
  ['trostky', 'trot', 'troskyite', 'troskyism'],
  ['usa', 'us', 'america', 'united states', 'united-states'],
  ['uyghur', 'uighur', 'xinjiang'],
  ['xi', 'jinping'],
  ['zapatista', 'ezln'],
]

export const addPlurals = (tags: string[]): string[] =>
  uniq(
    // @ts-ignore
    tags.map((tag) => [pluralize.plural(tag), pluralize.singular(tag)]).flat()
  )

export const addTagAliases = (tags: string[]): string[] => {
  const pluralized = addPlurals(tags)
  // @ts-ignore
  const possibleAliases = tagAliases
    .filter((xs) => xs.find((x) => pluralized.includes(x)))
    // @ts-ignore
    .flat()
  return uniq([...pluralized, ...possibleAliases])
}
