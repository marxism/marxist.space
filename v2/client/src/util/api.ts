import { getJson as _getJson } from 'fetchyeah'

const baseUrl =
  process.env.NODE_ENV === 'development'
    ? 'http://localhost:9090/api'
    : `https://${window.location.hostname}/api`

const getJson = (url) => _getJson(`${baseUrl}/${url}`)

export const getPosts = () => getJson('all')

export const getPost = (id) => getJson(`post/${id}`)

export const getTags = () => getJson('tags')
