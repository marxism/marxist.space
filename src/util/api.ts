import { getTags, getAll, fullTextSearch, filterByTags } from '../db'

export const fetchResources = () => Promise.resolve(getAll())

export const fetchTags = () => Promise.resolve(getTags())

export const searchResources = (text = '') =>
  Promise.resolve(fullTextSearch(text))

export const filterResources = (tags: Array<string> = []) =>
  Promise.resolve(filterByTags(tags))
