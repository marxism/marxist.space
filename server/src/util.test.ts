/* eslint-env jest */

import { uniq, addPlurals, addTagAliases } from './util'

describe('util', () => {
  test('uniq', () => {
    expect(uniq([1, 2, 1])).toMatchSnapshot()
    expect(uniq(['a', 'b'])).toMatchSnapshot()
  })

  test('addPlurals', () => {
    const tags = ['trot', 'library', 'documentary', 'people']
    expect(addPlurals(tags)).toMatchSnapshot()
  })

  test('addTagAliases', () => {
    const tags = ['abolition', 'israel', 'mao', 'essay']
    expect(addTagAliases(tags)).toMatchSnapshot()
  })
})
