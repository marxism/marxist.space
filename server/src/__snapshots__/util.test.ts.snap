// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util addPlurals 1`] = `
Array [
  "trots",
  "trot",
  "libraries",
  "library",
  "documentaries",
  "documentary",
  "people",
  "person",
]
`;

exports[`util addTagAliases 1`] = `
Array [
  "abolitions",
  "abolition",
  "israels",
  "israel",
  "maos",
  "mao",
  "essays",
  "essay",
  "police",
  "prison",
  "cop",
  "policing",
  "palestine",
  "zedong",
  "tsetung",
  "mlm",
  "maoist",
  "maoism",
]
`;

exports[`util uniq 1`] = `
Array [
  1,
  2,
]
`;

exports[`util uniq 2`] = `
Array [
  "a",
  "b",
]
`;
